[MKL-DNN] Pool & LRN Grad Ops NHWC support

jczaja opened the pull request:
This PR is introducing NHWC support for LRN and pool Grad ops of MKL-DNN as discussed at #20964 

**Note:**
There are changes in mean_op.cc and unused_var_check.cc that I would like @dzhwinter or @zhangting2020 or @sneaxiy to look at. Explanation is here:
https://github.com/PaddlePaddle/Paddle/issues/20964#issuecomment-565508163

sneaxiy commented on paddle/fluid/framework/operator.cc:
```diff
@@ -1180,12 +1178,38 @@ Scope* OperatorWithKernel::PrepareData(
-      Tensor out;
-      TransformData(expected_kernel_key, kernel_type_for_var, *tensor_in, &out);
-      SetTensorToVariable(*var, out, trans_var);
+      if (no_buffer_ins && no_buffer_ins->count(var_name_item.first) > 0) {
```
You cannot move `if (no_buffer_ins && no_buffer_ins->count(var_name_item.first) > 0)` here. 

In the design of `no_buffer_ins`, the tensor buffer of vars (i.e. the `holder_` field of Tensor) inside `no_buffer_ins` may be written (clear or written with other value) by other threads at the same time. Therefore, it is not thread safe if you read or write the tensor buffer of vars inside `no_buffer_vars`. Because the data transform does not handle these vars, the current implementation of Paddle is right when running ops in multi-threads. The only methods you can use are `Tensor::dims()`, `Tensor::lod()`, `Tensor::layout()`, `Tensor::offset()`, etc. Any Tensor methods which read or write `Tensor::holder_` are not allowed, for example `Tensor::IsInitialized()`, `Tensor::type()`, `Tensor::place()`, `Tensor::data()`, etc.

 In the current implementation of this PR, you check whether the tensor inside `no_buffer_ins ` is initialized , which may cause error in multi-threads.
https://github.com/PaddlePaddle/Paddle/blob/a400a2f37fc4d4c636171be66088641b3889d906/paddle/fluid/framework/operator.cc#L1121

jczaja replied:
Ok, So If I understand correctly when there is no_buffer_ins we should not call Tensor methods you listed. I have updated code (only operator.cc) Please tell me if this is fine approach.

jczaja commented:
@grygielski , @LeoZhao-Intel Could you please take a look at this PR ?

LeoZhao-BJ commented on paddle/fluid/framework/operator.cc:
```diff
@@ -1120,72 +1123,112 @@ Scope* OperatorWithKernel::PrepareData(
+                  << " in Operator " << type_;
+        }
+#endif
+      } else {
```
can you avoid adding else branch to make change more clear to reviewer? e.g.
if (should_skip_input == true)
{
  xxxx
  continue;
}

if (!tensor_in->IsInitialized()) {
          continue;
        }

jczaja replied:
Ok, will do that

grygielski commented on paddle/fluid/framework/operator.cc:
```diff
@@ -1102,10 +1106,9 @@ Scope* OperatorWithKernel::PrepareData(
-      VLOG(7) << "Skip scanning input " << var_name_item.first
-              << " in Operator " << type_;
-      continue;
+      should_skip_input = true;
```
Wouldn't be here more clear to just initialize should_skip_input with value of predicate inside if()?

jczaja replied:
Good catch! Will change that.

jczaja commented:
@LeoZhao-Intel , @grygielski I have updated PR with your suggestions and also added fix to newly introduced  MatchShapeToLayout(). Please continue your review

LeoZhao-BJ commented:
LGTM

luotao1 commented on paddle/fluid/framework/operator.cc:
```diff
@@ -1143,14 +1182,17 @@ Scope* OperatorWithKernel::PrepareData(
       // over the  running of operators.
-      // We use a thread_local cache to fix that issue, the key in the cache is
+      // We use a thread_local cache to fix that issue, the key in the cache
+      // is
```
line1186, line1193, line1220 have some formatting problem, please fix in next PR.

jczaja replied:
ok

luotao1 commented:
Please revert https://github.com/PaddlePaddle/Paddle/pull/21747/commits/a48ae6949047b38a660f2ddd45b66611d30cfaea back in next PR.
